File: TimeFlies/vc/HomeViewController.m
//
//  FirstViewController.m
//  testTabbarController
//
//  Created by 林伟池 on 15/9/17.
//  Copyright (c) 2015年 林伟池. All rights reserved.
//

#import "HomeViewController.h"
#import "RWKnobControl.h"
#import "HomeModel.h"
#import "TimeSelectController.h"
#import "UIViewControllerDelegate.h"
@interface HomeViewController ()
@end

#define const_select_time @"open_select_time_board"

@implementation HomeViewController
{
    NSTimer* timer;
    RWKnobControl* _knobControl;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Do any additional setup after loading the view, typically from a nib.

    [self viewInit];
    
    [[NSNotificationCenter defaultCenter] addObserverForName:@"HomeModelChange" object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
        [self viewInit];
    }];
    [self onTimer:nil];
    
    
    _knobControl = [[RWKnobControl alloc] initWithFrame:self.knobPlaceholder.bounds];
    [self.knobPlaceholder addSubview:_knobControl];
    
    _knobControl.lineWidth = 4.0;
    _knobControl.pointerLength = 8.0;
    self.view.tintColor = [UIColor lightGrayColor];
    
    [_knobControl addObserver:self forKeyPath:@"value" options:0 context:NULL];
    
    // Hooks up the knob control
    [_knobControl addTarget:self
                     action:@selector(handleValueChanged:)
           forControlEvents:UIControlEventValueChanged];

}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}



#pragma mark - view init

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self viewInit];
    [self addTimer];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    [timer invalidate];
}

- (void)viewInit
{
    self.myBirthDay.text = [NSString stringWithFormat:NSLocalizedString(@"今天是生命中的第%d天", nil), [[HomeModel instance] getDaysNow]];
    self.myDeadDay.text = [NSString stringWithFormat:NSLocalizedString(@"还有%d天", nil), [[HomeModel instance] getDaysLeft]];
    float per = [[HomeModel instance] getDaysLeft] / (365 * 70.0);
    self.valueSlider.value = (1 - per);
    [_knobControl setValue:(1 - per) animated:YES];
    
    
    NSDate* current = [[NSDate alloc] initWithTimeIntervalSinceNow:0];
    
    NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:NSLocalizedString(@"yyyy年MM月dd日", nil)];
    
    self.myDay.text = [dateFormatter stringFromDate:current];
}

- (void)addTimer
{
    timer = [NSTimer scheduledTimerWithTimeInterval:0.5 target:self selector:@selector(onTimer:) userInfo:self repeats:YES];
}

- (void)onTimer:(id)sender
{
    NSDate* current = [[NSDate alloc] initWithTimeIntervalSinceNow:0];
    
    NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
    
    [dateFormatter setDateFormat:@"hh:mm:ss"];
    self.mySecond.text = [dateFormatter stringFromDate:current];
}

#pragma mark - ui

-(IBAction)onSwipe:(UISwipeGestureRecognizer*)sender
{
    [self.tabBarController setSelectedIndex:1];
    [self.tabBarController tabBar:self.tabBarController.tabBar didSelectItem:self.tabBarController.tabBar.items[1]];
}

- (IBAction)handleValueChanged:(id)sender {
    if(sender == self.valueSlider) {
        _knobControl.value = self.valueSlider.value;
    } else if(sender == _knobControl) {
        self.valueSlider.value = _knobControl.value;
    }
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if(object == _knobControl && [keyPath isEqualToString:@"value"]) {
        self.myDeadDay.text = [NSString stringWithFormat:NSLocalizedString(@"还有%d天", nil), (long)(365 * 70 * (1- _knobControl.value))];
        NSDate* date = [[HomeModel instance] getBirthDate];
        date = [[NSDate alloc] initWithTimeInterval:((long)(365 * 70 * (_knobControl.value)) * 24 * 3600) sinceDate:date];
        
        NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
        [dateFormatter setDateFormat:NSLocalizedString(@"yyyy年MM月dd日", nil)];
        
        self.myDay.text = [dateFormatter stringFromDate:date];
    }
}

#pragma mark - delegate

- (void)dismissViewController:(TimeSelectController *)mcv
{
    [self dismissViewControllerAnimated:YES completion:nil];
}
#pragma mark - notify

@end

File: TimeFlies/vc/home/HomeViewController.swift
//
//  HomeViewController.swift
//  TimeFlies
//
//  Created by 林伟池 on 16/4/7.
//  Copyright © 2016年 林伟池. All rights reserved.
//

import Foundation

let lyCommentDefault = "ly"

class SwiftHomeViewController: UIViewController, ModalViewControllerDelegate {
    static let constSelectTime = "open_select_time_board"
    
    @IBOutlet var myBirthDay:UILabel?
    @IBOutlet var myDeadDay:UILabel?
    @IBOutlet var myDay:UILabel?
    @IBOutlet var mySecond:UILabel?
    @IBOutlet var myValueSlider:UISlider?
    @IBOutlet var myKnobPlaceholder:UIView?
    
    var myTimer:Timer?
    var myKnobControl:RWKnobControl?
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        viewInit()
        
        NotificationCenter.default.addObserver(forName: NSNotification.Name(rawValue: "HomeModalChange"), object: nil, queue: nil) { (notify) -> Void in
            self.viewInit()
        }
        onTimer("" as AnyObject)
        myKnobControl = RWKnobControl(frame: self.myKnobPlaceholder!.bounds)
        self.myKnobPlaceholder?.addSubview(myKnobControl!)
        
        myKnobControl?.lineWidth = 4.0
        myKnobControl?.pointerLength = 8.0
        
        view.tintColor = UIColor.lightGray
        
        myKnobControl?.addObserver(self, forKeyPath: "value", options: .new, context: nil)
        
        myKnobControl?.addTarget(self, action: #selector(SwiftHomeViewController.handleValueChanged(_:)), for: .valueChanged)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    func viewInit() {
        myBirthDay?.text = String(format:NSLocalizedString("今天是生命中的第%d天", comment: lyCommentDefault), HomeModel.instance().getDaysNow())
        
        myDeadDay?.text = String(format: NSLocalizedString("还有%d天", comment: lyCommentDefault), HomeModel.instance().getDaysLeft())
        let per:Float = Float(HomeModel.instance().getDaysLeft()) / (365 * 70.0)
        myValueSlider?.value = 1.0 - per
        myKnobControl?.setValue(CGFloat(1.0 - per), animated: true)

        let current = Date(timeIntervalSinceNow: 0)
        let dataFormatter = DateFormatter()
        dataFormatter.dateFormat = NSLocalizedString("yyyy年MM月dd日", comment: lyCommentDefault)
        myDay?.text = dataFormatter.string(from: current)
    }
    

    //MARK: view init

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        viewInit()
        addTimer()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        myTimer?.invalidate()
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
    }
    
    deinit {
        print("dealloc \(self)")
    }

    func addTimer() {
        myTimer = Timer.scheduledTimer(timeInterval: 0.5, target: self, selector: #selector(SwiftHomeViewController.onTimer(_:)), userInfo: self, repeats: true)
    }

    func onTimer(_ sender: AnyObject) {
        let current = Date(timeIntervalSinceNow: 0)
        let dataFormatter = DateFormatter()
        dataFormatter.dateFormat = "HH:mm:ss"
        mySecond?.text = dataFormatter.string(from: current)
    }
    
    //MARK: UI
    @IBAction func onSwipe(_ sender: UISwipeGestureRecognizer) {
        tabBarController?.selectedIndex = 1
        tabBarController?.tabBar((tabBarController?.tabBar)!, didSelect: (tabBarController?.tabBar.items![1])!)
    }
    
    @IBAction func onSwipeRight(_ sender: UISwipeGestureRecognizer) {
//        tabBarController?.selectedIndex = 3
    }
    
    @IBAction func handleValueChanged(_ sender: AnyObject) {
        if (sender as? NSObject) == self.myValueSlider {
            myKnobControl?.value = CGFloat((myValueSlider?.value)!)
        }
        if (sender as? NSObject) == self.myKnobControl {
            myValueSlider?.value = Float((myKnobControl?.value)!)
        }
    }

    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        
        if object as? NSObject == myKnobControl && keyPath! == "value" {
            myDeadDay?.text = String(format: NSLocalizedString("还有%d天", comment: lyCommentDefault), Int(365.0 * 70.0 * (1.0 - (myKnobControl?.value)!)))
            var date = HomeModel.instance().getBirthDate()
            date = Date(timeInterval: Double(365.0 * 70 * (myKnobControl?.value)! * 24 * 3600), since: date!)
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = NSLocalizedString("yyyy年MM月dd日", comment: lyCommentDefault)
            
            myDay?.text = dateFormatter.string(from: date!)
        }
    }

    //MARK: delegate
    func dismissViewController(_ mcv: TimeSelectController) {
        dismiss(animated: true) { () -> Void in
            
        }
    }
}
